nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+                    .whiteen()
+                    .clear_bit() // no data whitening
+            });
+
+            // CRC configuration required by the IEEE spec: x**16 + x**12 + x**5 + 1
+            radio
+                .radio
+                .crccnf
+                .write(|w| w.len().two().skipaddr().ieee802154());
+            radio.radio.crcpoly.write(|w| w.crcpoly().bits(0x11021));
+            radio.radio.crcinit.write(|w| w.crcinit().bits(0));
+
+            // permanent shortcuts
+            radio.radio.shorts.write(|w| w.txready_start().set_bit());
+        }
+
+        // set default settings
+        radio.set_cca(DEFAULT_CCA);
+        radio.set_channel(DEFAULT_CHANNEL);
+        radio.set_sfd(DEFAULT_SFD);
+        radio.set_txpower(DEFAULT_TXPOWER);
+
+        radio
+    }
+
+    /// Changes the radio channel
+    pub fn set_channel(&mut self, channel: Channel) {
+        self.disable();
+
+        // NOTE(unsafe) radio is currently disabled
+        unsafe {
+            self.radio
+                .frequency
+                .write(|w| w.map().clear_bit().frequency().bits(channel as u8))
+        }
+    }
+
+    /// Changes the Clear Channel Assessment method
+    pub fn set_cca(&mut self, cca: Cca) {
+        self.disable();
+
+        match cca {
+            Cca::CarrierSense => self.radio.ccactrl.write(|w| w.ccamode().carrier_mode()),
+        }
+    }
+
+    /// Changes the Start of Frame Delimiter
+    pub fn set_sfd(&mut self, sfd: u8) {
+        // FIXME don't completely turn off the radio; RXIDLE or TXIDLE are probably OK
+        self.disable();
+
+        self.radio.sfd.write(|w| unsafe { w.sfd().bits(sfd) });
+    }
+
+    /// Changes the TX power
+    pub fn set_txpower(&mut self, power: TxPower) {
+        // FIXME don't completely turn off the radio; RXIDLE or TXIDLE are probably OK
+        self.disable();
+
+        self.radio
+            .txpower
+            .write(|w| w.txpower().variant(power._into()));
+    }
+
+    /// Recevies one radio packet and copies its contents into the given `packet` buffer
+    ///
+    /// This methods returns the `Ok` variant if the CRC included the packet was successfully
+    /// validated by the hardware; otherwise it returns the `Err` variant. In either case, `packet`
+    /// will be updated with the received packet's data
+    pub fn recv(&mut self, packet: &mut Packet) -> Result<(), ()> {
+        // NOTE we do NOT check the address of `packet`; see comment in `Packet::new` for details
+
+        // go to the RXIDLE state
+        self.enable_rx();
+
+        // NOTE(unsafe) DMA transfer has not yet started
+        // set up RX buffer
+        unsafe {
+            self.radio
+                .packetptr
+                .write(|w| w.packetptr().bits(packet.buffer.as_mut_ptr() as u32));
+        }
+
+        // start transfer
+        dma_start_fence();
+        self.radio.tasks_start.write(|w| w.tasks_start().set_bit());
```

Comment:
That's exacly the place where I would clear `Event::End` flag, to be sure that it is not unintentionally set by some previously executed operation.